File: src/widgets/todolist.c
#include "todolist.h"
#include "../global.h"
#include "todo.h"

GtkWidget* TodoList()
{
    todos_list = adw_preferences_page_new();
    // Load todos as array [['todo1', 'sub1', 'sub2'], ['todo2', 'sub1', 'sub2']]
    GVariant* todos = g_settings_get_value(settings, "todos");
    // For each sub-array:
    for (int i = 0; i < g_variant_n_children(todos); i++) {
        // Get string array { 'todo1', 'sub1', 'sub2', NULL }
        const gchar** todo_items = g_variant_get_strv(g_variant_get_child_value(todos, i), NULL);
        // Add new todo row and pass array as arg
        adw_preferences_page_add(ADW_PREFERENCES_PAGE(todos_list), Todo((const gchar**)todo_items));
    }

    return todos_list;
}

File: src/widgets/todolist.py
from gi.repository import Adw
from ..globals import data
from ..data import UserData
from .todo import Todo


class TodoList(Adw.PreferencesPage):
    def __init__(self):
        super().__init__()
        data["todo_list"] = self
        self.load_todos()

    def load_todos(self):
        data = UserData.get()
        if data["todos"] != {}:
            for todo in data["todos"]:
                self.add(Todo(todo, data["todos"][todo]["sub"]))
